File: sdk/python/velocity_mcp_client/transport.py
"""Transport implementations for MCP client."""

import asyncio
import json
from abc import ABC, abstractmethod
from typing import Optional
import httpx

from .types import JsonRpcRequest, JsonRpcResponse
from .errors import TransportError, ConnectionClosedError


class Transport(ABC):
    """Abstract transport for MCP communication."""
    
    @abstractmethod
    async def send(self, request: JsonRpcRequest) -> JsonRpcResponse:
        """Send a JSON-RPC request and receive a response."""
        pass
    
    @abstractmethod
    async def close(self) -> None:
        """Close the transport connection."""
        pass


class HttpTransport(Transport):
    """HTTP transport - communicates via HTTP POST."""
    
    def __init__(self, url: str, api_key: Optional[str] = None, timeout: float = 30.0):
        """
        Initialize HTTP transport.
        
        Args:
            url: MCP server URL (e.g., "http://localhost:3000/mcp")
            api_key: Optional API key for authentication
            timeout: Request timeout in seconds
        """
        self.url = url
        self.api_key = api_key
        self.timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None
    
    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client."""
        if self._client is None or self._client.is_closed:
            headers = {}
            if self.api_key:
                headers["Authorization"] = f"Bearer {self.api_key}"
            
            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(self.timeout),
                headers=headers,
            )
        return self._client
    
    async def send(self, request: JsonRpcRequest) -> JsonRpcResponse:
        """Send HTTP request."""
        try:
            client = await self._get_client()
            response = await client.post(
                self.url,
                json=request.model_dump(by_alias=True, exclude_none=True),
            )
            response.raise_for_status()
            return JsonRpcResponse.model_validate(response.json())
        except httpx.TimeoutException as e:
            raise TransportError(f"Request timed out: {e}")
        except httpx.HTTPStatusError as e:
            raise TransportError(f"HTTP error {e.response.status_code}: {e.response.text}")
        except Exception as e:
            raise TransportError(f"Transport error: {e}")
    
    async def close(self) -> None:
        """Close HTTP client."""
        if self._client and not self._client.is_closed:
            await self._client.aclose()


class StdioTransport(Transport):
    """Stdio transport - communicates via stdin/stdout of a subprocess."""
    
    def __init__(self, command: str, args: Optional[list[str]] = None):
        """
        Initialize stdio transport.
        
        Args:
            command: Command to run (e.g., "velocity_mcp")
            args: Command arguments (e.g., ["--mode", "stdio"])
        """
        self.command = command
        self.args = args or []
        self._process: Optional[asyncio.subprocess.Process] = None
        self._request_id = 0
    
    async def _ensure_process(self) -> asyncio.subprocess.Process:
        """Ensure subprocess is running."""
        if self._process is None or self._process.returncode is not None:
            self._process = await asyncio.create_subprocess_exec(
                self.command,
                *self.args,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
        return self._process
    
    async def send(self, request: JsonRpcRequest) -> JsonRpcResponse:
        """Send request via stdio."""
        try:
            process = await self._ensure_process()
            
            if process.stdin is None or process.stdout is None:
                raise ConnectionClosedError("Process stdin/stdout not available")
            
            # Send request
            request_json = request.model_dump_json(by_alias=True, exclude_none=True)
            process.stdin.write((request_json + "\n").encode())
            await process.stdin.drain()
            
            # Read response
            response_line = await process.stdout.readline()
            if not response_line:
                raise ConnectionClosedError("Process closed stdout")
            
            response_json = json.loads(response_line.decode())
            return JsonRpcResponse.model_validate(response_json)
            
        except Exception as e:
            raise TransportError(f"Stdio transport error: {e}")
    
    async def close(self) -> None:
        """Close subprocess."""
        if self._process and self._process.returncode is None:
            self._process.terminate()
            try:
                await asyncio.wait_for(self._process.wait(), timeout=5.0)
            except asyncio.TimeoutError:
                self._process.kill()

File: client/src/transport.rs
//! Transport implementations for MCP client

use crate::error::{Error, Result};
use crate::types::{JsonRpcRequest, JsonRpcResponse};
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::Mutex;

/// Transport trait for MCP communication
#[async_trait::async_trait]
pub trait Transport: Send + Sync {
    /// Send a JSON-RPC request and receive a response
    async fn send(&self, request: JsonRpcRequest) -> Result<JsonRpcResponse>;
    
    /// Close the transport connection
    async fn close(&self) -> Result<()>;
}

/// Stdio transport - communicates via stdin/stdout
pub struct StdioTransport {
    child: Mutex<Child>,
}

impl StdioTransport {
    /// Create a new stdio transport by spawning a process
    pub fn new(command: &str, args: &[&str]) -> Result<Self> {
        let child = Command::new(command)
            .args(args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()?;
        
        Ok(Self {
            child: Mutex::new(child),
        })
    }
}

#[async_trait::async_trait]
impl Transport for StdioTransport {
    async fn send(&self, request: JsonRpcRequest) -> Result<JsonRpcResponse> {
        let mut child = self.child.lock().await;
        
        // Take stdin and stdout to avoid borrow conflicts
        let mut stdin = child.stdin.take().ok_or(Error::ConnectionClosed)?;
        let mut stdout = child.stdout.take().ok_or(Error::ConnectionClosed)?;
        
        // Serialize and send request
        let request_json = serde_json::to_string(&request)?;
        stdin.write_all(request_json.as_bytes()).await?;
        stdin.write_all(b"\n").await?;
        stdin.flush().await?;
        
        // Read response
        let mut reader = BufReader::new(&mut stdout);
        let mut response_line = String::new();
        reader.read_line(&mut response_line).await?;
        
        if response_line.is_empty() {
            return Err(Error::ConnectionClosed);
        }
        
        let response: JsonRpcResponse = serde_json::from_str(&response_line)?;
        
        // Put stdin and stdout back
        child.stdin = Some(stdin);
        child.stdout = Some(stdout);
        
        Ok(response)
    }
    
    async fn close(&self) -> Result<()> {
        let mut child = self.child.lock().await;
        child.kill().await?;
        Ok(())
    }
}

/// HTTP transport - communicates via HTTP/SSE
pub struct HttpTransport {
    client: reqwest::Client,
    url: String,
    api_key: Option<String>,
}

impl HttpTransport {
    /// Create a new HTTP transport
    pub fn new(url: &str, api_key: Option<String>) -> Result<Self> {
        let client = reqwest::Client::new();
        Ok(Self {
            client,
            url: url.to_string(),
            api_key,
        })
    }
}

#[async_trait::async_trait]
impl Transport for HttpTransport {
    async fn send(&self, request: JsonRpcRequest) -> Result<JsonRpcResponse> {
        let mut req_builder = self.client.post(&self.url).json(&request);
        
        if let Some(api_key) = &self.api_key {
            req_builder = req_builder.header("Authorization", format!("Bearer {}", api_key));
        }
        
        let response = req_builder.send().await?;
        
        if !response.status().is_success() {
            let status = response.status();
            let error_text = response.text().await.unwrap_or_default();
            return Err(Error::Server(format!("HTTP {}: {}", status, error_text)));
        }
        
        let rpc_response: JsonRpcResponse = response.json().await?;
        Ok(rpc_response)
    }
    
    async fn close(&self) -> Result<()> {
        // HTTP transport doesn't need explicit close
        Ok(())
    }
}
